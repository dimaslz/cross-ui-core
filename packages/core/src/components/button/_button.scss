@import '../../assets/css/_tailwind.scss';

$colors: 'primary', 'secondary', 'accent', 'danger', 'warning', 'success', 'info';
$colors-maps: (
  "primary": "blue",
  "secondary": "gray",
  "accent": "indigo",
  "danger": "red",
  "warning": "yellow",
  "success": "green",
  "info": "pink",
);

.CrossUIButton {
  @apply cursor-pointer border-solid font-sans flex justify-center;

  @each $color in $colors {
    &.#{$color} {
      $color-value: map-get($colors-maps, $color);

      @apply inline-flex items-center antialiased border-2 border-transparent leading-none text-sm px-4 bg-#{$color-value}-500 text-white font-semibold;
      @apply text-xs px-4;

      &.pill,
      &.circle {
        @apply rounded-full;
      }

      &.circle,
      &.square {
        @apply items-center justify-center p-1;
      }

      // SIZES
      &.small {
        @apply h-6 text-xs;

        &.circle,
        &.square {
          @apply w-6;
        }
      }

      &.medium {
        @apply h-8 text-sm;

        &.circle,
        &.square {
          @apply w-8;
        }
      }

      &.large {
        @apply h-10 text-base;

        &.circle,
        &.square {
          @apply w-10;
        }
      }

      &.simple {
        @apply outline-none bg-#{$color-value}-500;

        &:hover:not(:disabled) {
          @apply bg-#{$color-value}-600;
        }

        &:focus {
          @apply border-#{$color-value}-700;
        }
      }

      &.outline {
        @apply outline-none bg-transparent text-#{$color-value}-700 font-semibold border border-#{$color-value}-500;

        &:hover:not(:disabled) {
          @apply bg-#{$color-value}-600 text-white border-transparent;
        }

        &:focus {
          @apply border-#{$color-value}-700;
        }
      }

      &.elevated {
        @apply outline-none bg-white text-gray-800 font-semibold border border-gray-400 shadow;

        &:focus:not(:disabled),
        &:hover:not(:disabled) {
          @apply bg-gray-100;
        }
      }

      &:disabled {
        @apply opacity-50 cursor-not-allowed;
      }
    }
  }
}
