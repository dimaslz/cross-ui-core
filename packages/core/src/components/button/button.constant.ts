export const enum TYPE {
  BUTTON = 'button',
  SUBMIT = 'submit',
}
