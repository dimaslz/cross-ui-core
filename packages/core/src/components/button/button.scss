.CrossUIButton {
  @apply cursor-pointer border-solid font-sans flex justify-center;

  @each $color in $colors {
    &.#{$color} {
      $color-value: map-get($colors-map, $color);

      @apply inline-flex items-center antialiased border-2 border-transparent leading-none text-sm px-4 bg-#{$color-value}-500 text-white font-semibold;
      @apply text-xs px-4;

      &.pill,
      &.circle {
        @apply rounded-full;
      }

      &.circle,
      &.square {
        @apply items-center justify-center p-1;
      }

      // SIZES
      &.small {
        @apply h-6 text-xs;

        &.circle,
        &.square {
          @apply w-6;
        }
      }

      &.medium {
        @apply h-8 text-sm;

        &.circle,
        &.square {
          @apply w-8;
        }
      }

      &.large {
        @apply h-10 text-base;

        &.circle,
        &.square {
          @apply w-10;
        }
      }

      &.simple {
        @apply outline-none bg-#{$color-value}-500;

        &:hover:not(:disabled) {
          @apply bg-#{$color-value}-600;
        }

        &:active:not(:disabled) {
          @apply bg-#{$color-value}-900;
        }
      }

      &.outline {
        @apply outline-none bg-transparent text-#{$color-value}-700 font-semibold border border-#{$color-value}-700;

        svg {
          @apply text-#{$color-value}-700;
        }

        &:hover:not(:disabled) {
          @apply border-#{$color-value}-400 text-#{$color-value}-400 border-transparent;
        }

        &:active:not(:disabled) {
          @apply text-#{$color-value}-900 border-#{$color-value}-900;
        }
      }

      &.simple, &.outline {
        &:focus {
          @apply ring ring-gray-200;
        }
      }
    }

    &:disabled {
      @apply opacity-50 cursor-not-allowed;
    }
  }

  &--full-width {
    @apply w-full;
  }
}
