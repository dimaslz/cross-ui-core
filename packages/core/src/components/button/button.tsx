import { Component, h, Prop, Element, Host, Listen } from '@stencil/core';

@Component({
  tag: 'cross-button',
  // styleUrl: 'button.scss',
  // shadow: true,
})
export class Button {
  @Element() el: HTMLElement;
  @Prop({ mutable: true }) ui: string = 'simple';
  @Prop() color: string = 'primary';
  @Prop() size: string = 'medium';
  @Prop() circle: boolean = false;
  @Prop() square: boolean = false;
  @Prop() pill: boolean = false;
  @Prop() fullWidth: boolean = false;
  @Prop() disabled: boolean = false;
  @Prop() selected: boolean = false;
  @Prop() type: string = 'button';

  @Listen('click', { capture: true })
  onClickHandler($event) {
    if (this.disabled) {
      $event.stopPropagation();
      return false;
    }

    this?.el?.onclick?.call(this, $event);
  }

  render() {
    const extraClasses = this.el?.className
      .split(' ')
      .filter(c => c !== 'hydrated')
      .join(' ');
    const style: string = [
      'CrossUIButton',
      this.pill ? 'pill' : '',
      this.color,
      this.size,
      this.ui,
      this.circle ? 'circle' : '',
      this.square ? 'square' : '',
      this.fullWidth ? 'w-full' : '',
      extraClasses,
    ].join(' ');

    return (
      <Host>
        <button class={`${style}`} type={this.type} disabled={this.disabled}>
          <slot></slot>
        </button>
      </Host>
    );
  }
}
