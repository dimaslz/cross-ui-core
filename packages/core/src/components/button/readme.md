# cross-button

<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type      | Default     |
| ----------- | ------------ | ----------- | --------- | ----------- |
| `circle`    | `circle`     |             | `boolean` | `false`     |
| `color`     | `color`      |             | `string`  | `'primary'` |
| `disabled`  | `disabled`   |             | `boolean` | `false`     |
| `fullWidth` | `full-width` |             | `boolean` | `false`     |
| `outline`   | `outline`    |             | `boolean` | `false`     |
| `pill`      | `pill`       |             | `boolean` | `false`     |
| `selected`  | `selected`   |             | `boolean` | `false`     |
| `size`      | `size`       |             | `string`  | `'medium'`  |
| `square`    | `square`     |             | `boolean` | `false`     |
| `type`      | `type`       |             | `string`  | `'button'`  |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
