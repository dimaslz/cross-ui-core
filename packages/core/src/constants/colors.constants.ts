export enum COLOR {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
  ACCENT = 'accent',
  DANGER = 'danger',
  WARNING = 'warning',
  SUCCESS = 'success',
  INFO = 'info',
}

export const colors: string[] = Object.values(COLOR);
