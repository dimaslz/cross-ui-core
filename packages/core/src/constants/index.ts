export { colors } from './colors.constants';
export { sizes } from './sizes.constants';