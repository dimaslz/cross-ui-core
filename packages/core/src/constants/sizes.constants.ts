export enum SIZE {
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large',
}

export const sizes: string[] = Object.values(SIZE);
