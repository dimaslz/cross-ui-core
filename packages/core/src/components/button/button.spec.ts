// adds special assertions like toHaveTextContent
import {
  getByTestId,
  fireEvent,
} from '@testing-library/dom'
import '@testing-library/jest-dom/extend-expect'
import { newSpecPage } from '@stencil/core/testing';

import { colors, sizes } from '../../constants';

export * from './button';
import { Button } from './button';

async function initialSetup(attrs = [], options?) {
  const { testId = true } = options || { testId: true };

  const page = await newSpecPage({
    components: [Button],
    html: `<cross-button ${attrs.join(' ')} ${testId ? 'data-testid="test"' : ''}>Click me!</cross-button>`
  });

  return {
    page,
    button: testId && getByTestId(page.body, 'test'),
  } as any;
}

describe('cross-button', () => {
  it('builds', () => {
    expect(new Button()).toBeTruthy();
  });

  it('Snapshot', async () => {
    const { page } = await initialSetup([], { testId: false });

    expect(page.root).toMatchSnapshot();
  });

  it('disabled', async () => {
    const { button } = await initialSetup(['disabled']);

    expect(button.disabled).toBe(true);
  });

  it('full width', async () => {
    const { button } = await initialSetup(['full-width']);

    let classes = button.querySelector('button').className;

    expect(classes.includes('full-width')).toBe(true);
  });

  describe('Events', () => {
    it('click', async () => {
      const clickEvent = jest.fn();
      const { button } = await initialSetup();

      (button as any).onClick = clickEvent;
      button.click();

      expect(clickEvent).toBeCalled();
    });

    it('blur', async () => {
      const blurEvent = jest.fn();

      const { button } = await initialSetup();

      button.click();
      (button as any).onBlur = blurEvent;

      fireEvent.blur(button);

      expect(blurEvent).toBeCalled();
    });

    it('focus', async () => {
      const focusEvent = jest.fn();

      const { button } = await initialSetup();
      button.click();

      (button as any).onFocus = focusEvent;

      fireEvent.focus(button);

      expect(focusEvent).toBeCalled();
    });
  });

  describe('Colors', () => {
    it.each(
      colors.map(c => [c, c])
    )(`for color="%s" should have the class %s`, async (color) => {
      const { button } = await initialSetup([`color="${color}"`]);

      let classes = button.querySelector('button').className;

      expect(classes.includes(color)).toBe(true);
    });

    it('By default the color should be "primary"', async () => {
      const { button } = await initialSetup();

      const classes = button.querySelector('button').className;

      expect(classes.includes('primary')).toBe(true);
    });

    it('for color="unknown" color should throw an error', async () => {
      const color = 'unknown';

      try {
        await newSpecPage({
          components: [Button],
          html: `<cross-button color="${color}" data-testid="test">Click me!</cross-button>`
        });
      } catch (err) {
        expect(err.message).toBe(`Color "${color}" is not allowed. Please, use one of the following options: ${colors.join(', ')}`);
      }
    });
  });

  describe('UI\'s', () => {
    it.each(
      ['outline', 'square', 'circle', 'pill', 'active'].map(c => [c, c])
    )(`for attibute %s should have the class %s`, async (ui) => {
      const { button } = await initialSetup([ui]);

      const classes = button.querySelector('button').className;

      expect(classes.includes(ui)).toBe(true);
    });

    it('By default the UI should be "default"', async () => {
      const { button } = await initialSetup();

      const classes = button.querySelector('button').className;

      expect(classes.includes('default')).toBe(true);
    });
  });

  describe('Sizes', () => {
    it.each(
      sizes.map(c => [c, c])
    )(`for size="%s" should have the class %s`, async (size) => {
      const { button } = await initialSetup([`size="${size}"`]);

      let classes = button.querySelector('button').className;

      expect(classes.includes(size)).toBe(true);
    });

    it('By default the size should be "medium"', async () => {
      const { button } = await initialSetup();

      const classes = button.querySelector('button').className;

      expect(classes.includes('medium')).toBe(true);
    });

    it('for size="unknown" size should throw an error', async () => {
      try {
        await newSpecPage({
          components: [Button],
          html: `<cross-button size="unknown" data-testid="test">Click me!</cross-button>`
        });
      } catch (err) {
        expect(err.message).toBe(`Size "unknown" is not allowed. Please, use one of the following options: ${sizes.join(', ')}`);
      }
    });
  });
});
