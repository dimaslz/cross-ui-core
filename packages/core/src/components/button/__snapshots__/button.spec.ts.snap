// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cross-button Snapshot 1`] = `
<cross-button>
  <!---->
  <button class="CrossUIButton medium primary simple" type="button">
    Click me!
  </button>
</cross-button>
`;
